bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/handler/LivisiBridgeHandler.java

```diff
@@ -0,0 +1,991 @@
+import org.openhab.binding.livisismarthome.internal.discovery.LivisiDeviceDiscoveryService;
+import org.openhab.binding.livisismarthome.internal.listener.DeviceStatusListener;
+import org.openhab.binding.livisismarthome.internal.listener.EventListener;
+import org.openhab.binding.livisismarthome.internal.manager.DeviceStructureManager;
+import org.openhab.binding.livisismarthome.internal.manager.FullDeviceManager;
+import org.openhab.core.auth.client.oauth2.AccessTokenRefreshListener;
+import org.openhab.core.auth.client.oauth2.AccessTokenResponse;
+import org.openhab.core.auth.client.oauth2.OAuthClientService;
+import org.openhab.core.auth.client.oauth2.OAuthException;
+import org.openhab.core.auth.client.oauth2.OAuthFactory;
+import org.openhab.core.auth.client.oauth2.OAuthResponseException;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseBridgeHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link LivisiBridgeHandler} is responsible for handling the LIVISI SmartHome controller including the connection
+ * to the LIVISI SmartHome backend for all communications with the LIVISI SmartHome {@link DeviceDTO}s.
+ * <p/>
+ * It implements the {@link AccessTokenRefreshListener} to handle updates of the oauth2 tokens and the
+ * {@link EventListener} to handle {@link EventDTO}s, that are received by the {@link LivisiWebSocket}.
+ * <p/>
+ * The {@link DeviceDTO}s are organized by the {@link DeviceStructureManager}, which is also responsible for the
+ * connection
+ * to the LIVISI SmartHome webservice via the {@link LivisiClient}.
+ *
+ * @author Oliver Kuhl - Initial contribution
+ * @author Hilbrand Bouwkamp - Refactored to use openHAB http and oauth2 libraries
+ * @author Sven Strohschein - Renamed from Innogy to Livisi
+ */
+@NonNullByDefault
+public class LivisiBridgeHandler extends BaseBridgeHandler
+        implements AccessTokenRefreshListener, EventListener, DeviceStatusListener {
+
+    private final Logger logger = LoggerFactory.getLogger(LivisiBridgeHandler.class);
+    private final GsonOptional gson = new GsonOptional();
+    private final Object lock = new Object();
+    private final Map<String, DeviceStatusListener> deviceStatusListeners;
+    private final OAuthFactory oAuthFactory;
+    private final HttpClient httpClient;
+
+    private @Nullable LivisiClient client;
+    private @Nullable LivisiWebSocket webSocket;
+    private @Nullable DeviceStructureManager deviceStructMan;
```

Comment:
I had another look due to other comments, and it seems that `deviceStructMan` is guaranteed to be non-null after `initialize()`. So this could prevent all the `@Nullable`-madness:
```suggestion
    private @NonNullByDefault({}) DeviceStructureManager deviceStructMan;
```

Reply:
Ok, great. That trade-off makes a lot easier and shorter. I have now applied the same approach also for the "client" and "oAuthService" member variables within that class. These are also initialized within the initialize method.

Solved.